### ISSUE TITLE
Medusa will not start after upgrading python to 3.12.1 - AttributeError: module 'pkgutil'

### ISSUE THREAD

**Issue by rr0ss0rr** [pymedusa/Medusa#11633]
Just started to google this and seems to be an issue with `pkgutil.ImpImporter` and python 3.12

```
  File "/Applications/Medusa/lib/pkg_resources/__init__.py", line 2172, in <module>
    register_finder(pkgutil.ImpImporter, find_on_path)
                    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
```

Thanks

**Comment by John-Gee** [comment]
Any other option that downgrading?

**Comment by icfly2** [comment]
You need to add:

python -m ensurepip --upgrade
and 
python -m pip install --upgrade setuptools

You already have the second line, so just the first one should do the trick.

**Comment by p0psicles** [comment]
There are multiple issues that need to be resolved here.
* boto is deprecated
* appdirs is imported through pkg_resources.extern (pkg_resources used by babelfish used by guessit)
* .. wip

**Comment by RevillWeb** [comment]
Any update on when this might be fixed? Thank you.

**Comment by p0psicles** [comment]
no

**Comment by Linguista** [comment]
I just upgraded from Ubuntu 22.04 LTS to Ubuntu 24.04 LTS, which comes with Python 3.12.3. Medusa is completely broken in such setups. 

- You cannot downgrade your system Python without running the almost certain risk of destroying your OS installation. 

- The advice to add `python -m ensurepip --upgrade` produces an error:

`/usr/bin/python: No module named ensurepip`

- The advice to run `python -m pip install --upgrade setuptools' is broken, being blocked by design by the Python devs:

```
$ python -m pip install --upgrade setuptools
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification
```

- Installing Python 3.11 alongside 3.12 is possible, using the deadsnakes PPA, but Medusa doesn't like this either (though perhaps I'm not using the right commands/parameters):

```
$ /usr/bin/python3.11 /opt/medusa/start.py -q --nolaunch --datadir=/opt/medusa
ModuleNotFoundError: No module named '_cffi_backend'
thread '<unnamed>' panicked at /usr/share/cargo/registry/pyo3-0.20.2/src/err/mod.rs:788:5:
Python API call failed
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/opt/medusa/start.py", line 6, in <module>
    from medusa.__main__ import main
  File "/opt/medusa/medusa/__init__.py", line 24, in <module>
    initialize()
  File "/opt/medusa/medusa/init/__init__.py", line 37, in initialize
    _configure_guessit()
  File "/opt/medusa/medusa/init/__init__.py", line 202, in _configure_guessit
    from medusa.name_parser.guessit_parser import (
  File "/opt/medusa/medusa/name_parser/guessit_parser.py", line 12, in <module>
    from medusa.logger.adapters.style import BraceAdapter
  File "/opt/medusa/medusa/logger/__init__.py", line 53, in <module>
    import subliminal
  File "/opt/medusa/ext/subliminal/__init__.py", line 11, in <module>
    from .core import (AsyncProviderPool, ProviderPool, check_video, download_best_subtitles, download_subtitles,
  File "/opt/medusa/ext/subliminal/core.py", line 13, in <module>
    from rarfile import BadRarFile, NotRarFile, RarCannotExec, RarFile, Error, is_rarfile
  File "/opt/medusa/ext/rarfile.py", line 109, in <module>
    from cryptography.hazmat.primitives.ciphers import algorithms, modes, Cipher
  File "/usr/lib/python3/dist-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
    from cryptography.hazmat.primitives.ciphers.base import (
  File "/usr/lib/python3/dist-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
    from cryptography.exceptions import (
  File "/usr/lib/python3/dist-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
pyo3_runtime.PanicException: Python API call failed
```

In short, Medusa's use of a long-deprecated and now eliminated class makes it impossible to use on Ubuntu/Debian, and possibly other Linux distros. If there's a working workaround, do tell, as I haven't been able to find one. I'd rather stick with Medusa than restart my collection from scratch with Sonarr, but this issue has been open for five months now and shows no signs of being of interest to anyone.

**Comment by StSaens** [comment]
I just upgraded my Ubuntu install to latest LTS and was wondering why Medusa didn’t work anymore.

**Comment by Linguista** [comment]
To answer my own question, a workaround for this problem is to run Medusa in a Python virtual environment built on Python 3.11. I used `venv` in this case.

**Comment by StSaens** [comment]
> To answer my own question, a workaround for this problem is to run Medusa in a Python virtual environment built on Python 3.11. I used `venv` in this case.

Can you write down the steps/commands to get it running like that? I'm assuming the broken install can be migrated to such a virtual environment setup?

**Comment by Linguista** [comment]
> > To answer my own question, a workaround for this problem is to run Medusa in a Python virtual environment built on Python 3.11. I used `venv` in this case.
> 
> Can you write down the steps/commands to get it running like that? I'm assuming the broken install can be migrated to such a virtual environment setup?

I didn't document the process, unfortunately, but from memory and combing through my bash history, the steps are these:

**1. Install Python 3.11**
```
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt update
sudo apt install python3.11-full python3-venv python3.11-dev python3.11-distutils
```
_The last two packages may not be necessary, but I installed them anyway._


**2. Create the virtual environment**
```
mkdir -p ~/bin/py311
/usr/bin/python3.11 -m venv ~/bin/py311/
```

**3. Make sure Python 3.11 venv is installed and working**
```
source ~/bin/py311/bin/activate
python -V
```
_If all is well, Python should be version 3.11 in the venv._

**4. Install Medusa in the virtual environment and then install its dependencies**
```
cd ~/bin/py311
git clone https://github.com/pymedusa/Medusa.git
cd medusa
pip install -r requirements.txt
```

**5. Run Medusa**
`/home/YourUsername/bin/py311/bin/python3 /home/YourUsername/bin/py311/medusa/start.py -q --nolaunch --datadir=/home/YourUsername/bin/py311/medusa`

_I was extra cautious here, using full, absolute paths for everything. That may or may not be necessary, but the line above has proven to work in my systemd `medusa.service` file._

Hope this helps. But do keep in mind that it's quite possible I forgot some step or another.

**Comment by bob9931** [comment]
Still not working. Did a fresh git clone on Ubuntu 24.04, starting medusa gives
_"AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?"_

**Comment by snigel** [comment]
Just tried a fresh clone on Ubuntu 24.04.1 LTS (Python 3.12.3) and getting the same results.
`AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?`

```
Traceback (most recent call last):
  File "/home/snigel/src/Medusa/./start.py", line 6, in <module>
    from medusa.__main__ import main
  File "/home/snigel/src/Medusa/medusa/__init__.py", line 24, in <module>
    initialize()
  File "/home/snigel/src/Medusa/medusa/init/__init__.py", line 37, in initialize
    _configure_guessit()
  File "/home/snigel/src/Medusa/medusa/init/__init__.py", line 201, in _configure_guessit
    import guessit
  File "/home/snigel/src/Medusa/ext/guessit/__init__.py", line 8, in <module>
    from .api import guessit, GuessItApi
  File "/home/snigel/src/Medusa/ext/guessit/api.py", line 17, in <module>
    from .rules import rebulk_builder
  File "/home/snigel/src/Medusa/ext/guessit/rules/__init__.py", line 11, in <module>
    from .properties.episodes import episodes
  File "/home/snigel/src/Medusa/ext/guessit/rules/properties/episodes.py", line 16, in <module>
    from .title import TitleFromPosition
  File "/home/snigel/src/Medusa/ext/guessit/rules/properties/title.py", line 11, in <module>
    from .language import (
  File "/home/snigel/src/Medusa/ext/guessit/rules/properties/language.py", line 10, in <module>
    import babelfish
  File "/home/snigel/src/Medusa/ext/babelfish/__init__.py", line 14, in <module>
    from .converters import (LanguageConverter, LanguageReverseConverter, LanguageEquivalenceConverter, CountryConverter,
  File "/home/snigel/src/Medusa/ext/babelfish/converters/__init__.py", line 5, in <module>
    from pkg_resources import iter_entry_points, EntryPoint
  File "/home/snigel/src/Medusa/lib/pkg_resources/__init__.py", line 2172, in <module>
    register_finder(pkgutil.ImpImporter, find_on_path)
                    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?

```

**Comment by medariox** [comment]
Python 3.12 added with the new release.
